packages/big-design/src/components/Tree/TreeNode/TreeNode.tsx

```diff
@@ -0,0 +1,328 @@
+  StyledGap,
+  StyledLi,
+  StyledSelectableWrapper,
+  StyledText,
+} from './styled';
+
+interface PrivateTreeItemProps<T> {
+  state: TreeState<T>;
+  dispatch: Dispatch<Action<T>>;
+  iconless?: boolean;
+  selectable: TreeProps<T>['selectable'];
+  onExpand?: TreeProps<T>['onExpand'];
+  onCollapse?: TreeProps<T>['onCollapse'];
+  onSelect?: TreeProps<T>['onSelect'];
+}
+
+const flexItemProps: FlexItemProps = { flexShrink: 0, marginLeft: 'xxSmall' };
+
+const InternalTreeNode = <T extends unknown>({
+  children,
+  disabled,
+  dispatch,
+  icon,
+  iconless,
+  id,
+  label,
+  onCollapse,
+  onExpand,
+  onSelect,
+  selectable,
+  state,
+  value,
+}: TreeNodeProps<T> & PrivateTreeItemProps<T>): React.ReactElement<TreeNodeProps<T>> => {
+  const thisRef = useRef<TreeNodeRef<T>>({ children });
+  const nodeRef = useRef<HTMLLIElement | null>(null);
+  const selectableRef = useRef<HTMLLabelElement | null>(null);
+  const [isLoading, setIsLoading] = useState(false);
+  const expanded = useIsExpanded(state, id);
+  const selected = useIsSelected(state, value);
+
+  // Could be multiple elements in which are clicked.
+  // Typing to generic Element type since all other elements extend from it.
+  const handleNodeToggle = async (e?: React.MouseEvent<Element>) => {
+    dispatch({ type: 'FOCUS', id });
+
+    // Prevents the collapse/expand when clicking on a radio or checkbox
+    // Checks to see if every element inside the selectableRef gets clicked.
+    if (
+      (e?.target instanceof Node && selectableRef.current?.contains(e?.target)) ||
+      thisRef.current.children === undefined
+    ) {
+      return;
+    }
+
+    dispatch({ type: 'TOGGLE_NODE', id });
+
+    if (expanded) {
+      if (typeof onCollapse === 'function') {
+        const callbackValue = await onCollapse({
+          children: thisRef.current.children,
+          disabled,
+          expanded,
+          id,
+          label,
+          selected,
+          value,
+        });
+
+        if (callbackValue) {
+          thisRef.current = callbackValue;
+        }
+      }
+    } else {
+      if (typeof onExpand === 'function') {
+        setIsLoading(true);
+
+        const callbackValue = await onExpand({
+          children: thisRef.current.children,
+          disabled,
+          expanded,
+          id,
+          label,
+          selected,
+          value,
+        });
+
+        if (callbackValue) {
+          thisRef.current = callbackValue;
+        }
+
+        setIsLoading(false);
+      }
+    }
+  };
+
+  const handleNodeSelected = useCallback(() => {
+    if (selectable === undefined || value === undefined || disabled) {
+      return;
+    }
+
+    let newSelectedValues: T[] = [];
+
+    if (selectable === 'multi') {
+      if (state.selectedValues.includes(value)) {
+        newSelectedValues = state.selectedValues.filter((selectedValue) => selectedValue !== value);
+      } else {
+        newSelectedValues = [...state.selectedValues, value];
```

Comment:
Pretty much most of the state values can be changed to using a `Set`. Honestly, I probably should've done it initially. I'll go ahead and start convert over since I'm getting into cleanup and performance.